tasks/test_add_picklist_values.py

```diff
@@ -0,0 +1,67 @@
+import click
```

Comment:
@jstvz Soo... I've started to implement tests for the custom task but I'm a bit blocked by lack of knowledge. A little bit of help would get me going faster, I think.

I sort of have the mocks working... though I'm pretty sure when I run `python -m unittest` and it runs this test, it's actually doing a callout to example.com which we obviously don't want. 😆 

So I just need to figure out how to mock the results to get back what I want (there are multiple queries, callouts, and a deployment that's done as part of this task...)

If you could give me some guidance tomorrow or Wednesday that'd be great.

Reply:
Right.

## Running tests

We've switched to [pytest](https://docs.pytest.org/en/latest/contents.html) as our testing framework/test runner. If you've got a development virtualenv set up, you can run these tests with `pytest tasks/`.

## Mocking APIs

The [`CreateCommunity`](https://github.com/sfdo-tooling/CumulusCI/blob/master/cumulusci/tasks/salesforce/CreateCommunity.py) task began life as a custom task written by @dmreed, so it's a pretty good example. Its associated test, [`test_CreateCommunity.py`](https://github.com/sfdo-tooling/CumulusCI/blob/master/cumulusci/tasks/salesforce/tests/test_CreateCommunity.py) shows how to mock a series of API calls using the [responses](https://github.com/getsentry/responses) library. If we examine its first test, [`test_creates_community`](https://github.com/sfdo-tooling/CumulusCI/blob/master/cumulusci/tasks/salesforce/tests/test_CreateCommunity.py#L26-L66), we see that it includes this import:

``` python
from cumulusci.tasks.salesforce.tests.util import create_task
```

`create_task` handles setting up the project/org/task configs, and we see it called as the first line of the test:

``` python
@responses.activate
def test_creates_community(self):
    cc_task = create_task(CreateCommunity, task_options)
```

The `@responses.activate` decorator allows us to mock a series of HTTP responses. We define the URLs to be mocked:

``` python
servlet_url = "{}/sites/servlet.SitePrerequisiteServlet".format(
    cc_task.org_config.instance_url
)
community_url = "{}/services/data/v46.0/connect/communities".format(
    cc_task.org_config.instance_url
)
```

The first call is to [`org_config.start_url`](https://github.com/sfdo-tooling/CumulusCI/blob/master/cumulusci/core/config/OrgConfig.py#L89-L95), the others to the URLs defined above.

``` python
responses.add(
  method=responses.GET, url=cc_task.org_config.start_url, status=200
)
responses.add(method=responses.GET, url=servlet_url, status=200)
responses.add(method=responses.POST, url=community_url, status=200, json={})
responses.add(
  method=responses.GET,
  url=community_url,
  status=200,
  json={"communities": [{"name": "Test Community", "id": "000000000000000"}]},
)
```

The first two return no JSON payloads, but the second two do. This is where we'd place our expected returns from the tooling API.

The next line calls the task:

``` python
cc_task()
```

We then assert that:

The correct number of calls was made:

``` python
self.assertEqual(4, len(responses.calls))
```

The `start_url`, `servlet_url`, and `community_url` URLs were called in the correct order:

``` python
self.assertEqual(cc_task.org_config.start_url, responses.calls[0].request.url)
self.assertEqual(servlet_url, responses.calls[1].request.url)
self.assertEqual(community_url, responses.calls[2].request.url)
self.assertEqual(community_url, responses.calls[3].request.url)
```

The POST request body is what we expected:

``` python
self.assertEqual(
    json.dumps(
        {
            "name": "Test Community",
            "description": "Community Details",
            "templateName": "VF Template",
            "urlPathPrefix": "test",
        }
    ),
    responses.calls[2].request.body,
)
```

---
Let me know if you want to chat, @robertbwatson.

Reply:
@jstvz Going well so far. My test needs to know the API version of the Tooling API. Apparently it is using API v47.0. How do I get to that in my test so that the tests don't fail when the version of the Tooling API is updated?

Reply:
@robertbwatson I would specify `api_version = "47.0"` in the actual task implementation, to make sure that it doesn't suddenly start behaving differently if the API changes in a future API version. Then you can hardcode 47 in the test as well and not worry about it changing until you decide to.

Reply:
@davisagli Simple enough, thanks!